daemon/src/main.rs

```diff
@@ -0,0 +1,44 @@
+pub mod service;
+
+use std::{future::pending, sync::OnceLock};
+
+use oo7::portal::Keyring;
+use zbus::{ConnectionBuilder, Result};
+
+const SECRET_SERVICE_OBJECTPATH: &str = "/org/freedesktop/secrets_";
+// const SECRET_COLLECTION_OBJECTPATH: &str =
+// "/org/freedesktop/secrets_/collection"; const SECRET_SESSION_OBJECTPATH: &str
+// = "/org/freedesktop/secrets_/session";
+
+use crate::service::service::Service;
+
+pub static KEYRING: OnceLock<Keyring> = OnceLock::new();
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let secret_service = Service {
+        collections: Vec::new(),
+    };
+
+    // let collection = Collection::default();
+    // let session = Session {
+    // path: ObjectPath::default().into(),
+    // };
+
+    let _service = ConnectionBuilder::session()?
+        .name("org.freedesktop.secrets_")?
+        .serve_at(SECRET_SERVICE_OBJECTPATH, secret_service)?
+        //.serve_at(SECRET_COLLECTION_OBJECTPATH, collection)?
+        //.serve_at(SECRET_SESSION_OBJECTPATH, session)?
+        .build()
+        .await?;
+
+    let _keyring = match Keyring::load_default().await {
+        Ok(keyring) => KEYRING.set(keyring),
+        Err(_) => todo!(), // call some init to create default keyring
+    };
```

Comment:
All of this should be done inside `Service::new()` except the `serve_at` bit, which you would do at a later stage by calling `service.run().await?;`

Reply:
By doing so, the KEYRING, is something that no longer would be a Lazy lock, instead you would add it as a field to the `Service`, like `Arc<Keyring>` and pass it down the line to the collections or items if they need access to it.

Reply:
> All of this should be done inside `Service::new()` except the `serve_at` bit, which you would do at a later stage by calling `service.run().await?;`

Hey,

I have implemented the `Service::new()`: https://github.com/bilelmoussaoui/oo7/blob/10c73b94103ff0b9fb592bf0a3a765ceee9804fc/daemon/src/service/service.rs#L199 To implement the `service.run()` I have included `connection: ConnectionBuilder` field in `Service` struct. To this point everything worked well. But, `serve_at` requires the ownership of `service` (or whatever we pass it). So, I think `run()` can't be implemented as a instance method :)

> By doing so, the KEYRING, is something that no longer would be a Lazy lock, instead you would add it as a field to the `Service`, like `Arc<Keyring>` and pass it down the line to the collections or items if they need access to it.

What's the best approach to pass down `service`'s `keyring: Arc<Keyring>`. The only approach I can think and tried (it didn't go well) is use the return value of `service.keyring()` as `OnceLock` :)

Reply:
> To implement the service.run() I have included connection: ConnectionBuilder field in Service struct

You don't need to, you should setup the ConnectionBuilder inside `service.run()` + make the function take `self` and not `&self` and serve_at would work by passing it `self`

Reply:
With this change: https://github.com/bilelmoussaoui/oo7/blob/004f98a69f98456d86083b0b5f9264c4471c6c1e/daemon/src/service/service.rs#L212 
oo7-daemon is no longer responds to `busctl`. I usually run `busctl --user tree org.freedesktop.secrets_` after a major change. It's no longer there. I'm not sure what's causing this :)

Reply:
I am not sure what would cause that, the code changes looks sane to me.

Reply:
yeah, I agree :) I'll look into this again.